frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskLogs.tsx

```diff
@@ -0,0 +1,155 @@
+import * as React from 'react';
+import * as _ from 'lodash';
+import { K8sKind } from '@console/internal/module/k8s';
+import {
+  ResourceLog,
+  LOG_SOURCE_TERMINATED,
+  LoadingInline,
+} from '@console/internal/components/utils';
+import './PipelineTaskLogs.scss';
+import { containerToLogSourceStatus } from '../../utils/pipeline-utils';
+
+interface TektonLogProps {
+  obj?: Pod;
+  taskName: string;
+}
+
+interface Pod {
+  data: K8sKind;
+}
+interface TektonLogState {
+  renderContainers: any;
+  fetching: boolean;
+  targetHeight: number;
+}
+
+class TektonLogs extends React.Component<TektonLogProps, TektonLogState> {
+  scrollPane: React.RefObject<HTMLDivElement> = React.createRef();
+
+  scrollUntouched: boolean = true;
+
+  isAutoScrolling: boolean = false;
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      renderContainers: [],
+      fetching: false,
+      targetHeight: null,
+    };
+  }
+
+  componentDidMount() {
+    const containers = _.get(this.props, ['obj', 'data', 'spec', 'containers'], []);
+    const containerStatus = _.get(this.props, ['obj', 'data', 'status', 'containerStatuses'], []);
+    const firstrunningCont = containerStatus.findIndex(
+      (container) => containerToLogSourceStatus(container) !== LOG_SOURCE_TERMINATED,
+    );
+    const renderContainers = containers.slice(
+      0,
+      firstrunningCont === -1 ? containers.length : firstrunningCont + 1,
+    );
+
+    this.setState({ renderContainers });
+    firstrunningCont === -1
+      ? this.setState({ fetching: false })
+      : this.setState({ fetching: true });
+    this.scrollToBottom();
+    window.addEventListener('resize', this.sizeContainer, { passive: true });
+    this.sizeContainer();
+  }
+
+  componentWillReceiveProps(nextProps) {
+    const containers = _.get(nextProps, ['obj', 'data', 'spec', 'containers'], []);
+    const containerStatus = _.get(nextProps, ['obj', 'data', 'status', 'containerStatuses'], []);
+    if (this.props !== nextProps) {
+      const firstrunningCont = containerStatus.findIndex(
+        (container) => containerToLogSourceStatus(container) !== LOG_SOURCE_TERMINATED,
+      );
+      const renderContainers = containers.slice(
+        0,
+        firstrunningCont === -1 ? containers.length : firstrunningCont + 1,
+      );
+      this.setState({ renderContainers });
+      firstrunningCont === -1
+        ? this.setState({ fetching: false })
+        : this.setState({ fetching: true });
+    }
+    if (
+      _.get(this.props, ['obj', 'data', 'metadata', 'name'], false) !==
+      _.get(nextProps, ['obj', 'data', 'metadata', 'name'], false)
+    ) {
+      this.scrollUntouched = true;
+    }
+  }
+
+  componentDidUpdate() {
+    setTimeout(() => this.scrollToBottom(), 800); // buffer time for children to render
+  }
+
```

Comment:
Okay, so I feel the use of the lifecycle methods here can be improved. Firstly I think we should avoid using `componentWillReceiveProps` in any new components since it is going to be deprecated and also it has the potential of creating a lot of bugs if not handled properly. See if it is possible to just use `componentDidUpdate` in its place. Also `componentDidMount` has the same code as `componentWillReceiveProps` I am not sure if that is the only way to achieve what you are trying to do here. And if I am missing something or this is the only way to handle the lifecycle methods here please let me know.

Reply:
Yes, I do agree, I could've bettered out the usage of component life cycles. However, keeping in mind that this is part of an urgent requirement and the PR needs to go in real quick and also that it is currently only a workaround to cause least possible impact ( post 4.2 the whole implementation will be dropped and the `ResourceLog` component will be updated instead.

1. Use of `componentWillRecieveProps`, I need to use a pre-render method, using `componentDidUpdate` with `setState` will trigger the rendering twice. Using `componentWillUpdate` which is not very stable will be invoked on `stateChanges` as well, which is not the target trigger for the logic.  
2. The setting of state in `componentDidMount` and `componentWillRecieve` is the same however the `state` is derived out of different target data and there is some conditional logic in place.
Additionally there is addition of an eventlistener in `didMount` and there is resetting of the `scrollUntouched` flag in case the current Task is changed.